### PR TITLE
[SC-320] - Implement Export process in scrowl

### PR DESCRIPTION
**Pull request by SezarGantousOSG**
### Short summary

- For now we are using the fork from [this repo](https://github.com/andre-dietrich/simple-scorm-packager)
   - We can create our own fork in a later step after using it a bit more and see a need to extend 
- Created an example button in the renderer that will invoke the export process in the main process
-

---

### Test steps

Before starting please have the LMS going locally: https://www.notion.so/LMS-Local-Setup-ab83bca28bd04d61947a4a54e55b5c4b

---

NOTE: You cannot export while in docker so you will need to test this on host machine (permission conflict between the export process which needs root and electron which cannot start if the container has root access )

Once the app is up locally on host machine:

1. Create a folder called `res` in `course-authoring/src/scormBuild`
1. In that folder create a file called `start.html`
2. In `start.html` file add the following:
```
<!DOCTYPE html>
<html>
<head>
    <meta charset='utf-8'>
    <meta http-equiv='X-UA-Compatible' content='IE=edge'>
    <title>Page Title</title>
    <meta name='viewport' content='width=device-width, initial-scale=1'>
</head>
<body>
    <h1>Helo world!</h1>
</body>
</html>
```
3. Click the `export` button in the electron app and you should see a new zip folder in `src/scormPackages`
4. Insure that the zip file works by uploading it into a local LMS instance:
   1. Please follow these Instructions: https://www.notion.so/Test-SCORM-Package-041b79520b5042459d1cddc7208255a5

### CODE REVIEW

File: apps/course-authoring/src/main/index.ts [hunk 1]
```diff
@@ -20,6 +21,32 @@ ipcMain.on('ipc-example', async (event, arg) => {
   event.reply('ipc-example', msgTemplate('pong'));
 });
 
+// Renderer to Main and back to Renderer
+ipcMain.handle('startScormExportProcess', (event, args) => {
+  console.log(`SCORM package export of ${args} has started`);
+  
+  const scormProcess= buildPackage({
```
Comment (SezarGantousOSG) [inline apps/course-authoring/src/main/index.ts:28]: This is where the `magic` happens

File: apps/course-authoring/src/main/preload.ts [hunk 2]
```diff
@@ -27,3 +27,10 @@ contextBridge.exposeInMainWorld('electron', {
     },
   },
 });
+
+const exportScormApi = {
```
Comment (CiceroFonsecaOSG) [inline apps/course-authoring/src/main/preload.ts:31]: We already have the `electron` object being exposed to the main world @3 so I think we should get rid of this `exportScormApi` and the `contextBridge` below and add a more generic `invoke` method to the `ipcRenderer` object in the `contextBridge.exposeInMainWorld('electron', something like:
```js
    invoke(channel: string, ...args: unknown[]) {
      const validChannels = ['ipc-example', 'start-scorm-export-process'];
      if (validChannels.includes(channel)) {
        return ipcRenderer.invoke(channel, ...args);
      }
    },
```
This way we can use it in the renderer process like this:
```js
          onClick={async () => {
            const messageFromMain = await window.electron.ipcRenderer.invoke(
              'start-scorm-export-process',
              { courseName: 'Test Course' }
            );

            console.log(messageFromMain);
          }}
```
And get the same results without using the `contextBridge` twice and attaching two different new objects to the window.

File: apps/course-authoring/src/renderer/pages/home/index.tsx [hunk 3]
```diff
@@ -28,6 +28,22 @@ const Header = (
           Open
         </Btn>
       </div>
+      <div>
+        <Btn
```
Comment (CiceroFonsecaOSG) [inline apps/course-authoring/src/renderer/pages/home/index.tsx:32]: If we add the `invoke` method to the existing `window.electron` we can replace it with:
```tsx
        <Btn
          size="Sm"
          onClick={async () => {
            const messageFromMain = await window.electron.ipcRenderer.invoke(
              'start-scorm-export-process',
              { courseName: 'Test Course' }
            );

            console.log(messageFromMain);
          }}
        >
```

File: apps/course-authoring/types.d.ts [hunk 4]
```diff
@@ -1 +1,6 @@
-declare module '*.scss'
\ No newline at end of file
+declare module '*.scss'
```
Comment (CiceroFonsecaOSG) [inline apps/course-authoring/types.d.ts:1]: I think this `types.d.ts` is "too global", we could move it to the same folder as the `index.ts` file `/src/main` and call it `index.types.ts` for consistency.

Also, as we'll be getting rid of the `exportScormApi` and replacing it with a new method in the `window.electron.ipcrender` we can create something more generic for the type, kinda like:
```ts
declare global {
  interface Window {
    electron: {
      ipcRenderer: {
        [key: string]: (...args: unknown[]) => void;
      };
    };
  }
}

export {};
```

You'll notice TS angry at you when importing styles, I'm pretty sure it's been fixed in the new owlUI version that will be released soon.

**Review by CiceroFonsecaOSG — CHANGES_REQUESTED** [review]

**Comment by aubeOSG** [comment]
I was able to create the course ZIP file and upload it to the LMS (running locally) but the import failed. There are no error logs in the browser console.
Screen Shot 2022-04-20 at 1 54 48 PM


Looking at the logs from the LMS, there doesn't seem to be any error being messaged out.
Screen Shot 2022-04-20 at 1 53 48 PM

Before merging this in, we should have a course working in the LMS.

**Comment by SezarGantousOSG** [comment]
@CiceroFonsecaOSG made the changes. as for the types I just followed what electron had: https://www.electronjs.org/docs/latest/tutorial/context-isolation#usage-with-typescript

**Comment by aubeOSG** [comment]
With the current set of instructions:

> 1. Create a folder called res in course-authoring/src
> 2. In that folder create a file called start.html

And the packager config:

```json
{
  "startingPage": "/res/start.html",
  "source": "./src/scormBuild"
}
```

The scorm package will not load correctly into the LMS.

Files related to the course content need to be placed inside of the `source` folder as the files within the `source` are zipped together.

Additionally, I was not able to get the LMS to load the zip file correctly. I needed to run this command `docker-compose exec api sidekiq -C config/sidekiq.yml` in order to process the scorm package. The [local LMS setup instructions](https://www.notion.so/LMS-Local-Setup-ab83bca28bd04d61947a4a54e55b5c4b) should be updated with this command.

**Comment by SezarGantousOSG** [comment]
> With the current set of instructions:
> 
> > 1. Create a folder called res in course-authoring/src
> > 2. In that folder create a file called start.html
> 
> And the packager config:
> 
> ```json
> {
>   "startingPage": "/res/start.html",
>   "source": "./src/scormBuild"
> }
> ```
> 
> The scorm package will not load correctly into the LMS.
> 
> Files related to the course content need to be placed inside of the `source` folder as the files within the `source` are zipped together.
> 
> Additionally, I was not able to get the LMS to load the zip file correctly. I needed to run this command `docker-compose exec api sidekiq -C config/sidekiq.yml` in order to process the scorm package. The [local LMS setup instructions](https://www.notion.so/LMS-Local-Setup-ab83bca28bd04d61947a4a54e55b5c4b) should be updated with this command.

Glad you got it working.

1. You are correct, I forgot to add the `scormBuild` into the path for the first step. Sorry about that, I have added that in for what it's worth.
2. The LMS did add a layer for this task (took a day with a lot of different tries) and I honestly do not remember doing that command. Only this `docker exec osg_online_api bash sidekiq.sh` which is step 3 of the [LMS Setup instructions](https://www.notion.so/LMS-Local-Setup-ab83bca28bd04d61947a4a54e55b5c4b#55dfcdd3fc1a40d6abf6af99dc5c1380). The LMS seems to process fine without it for me, very strange. We can add that command though into the instructions, would like to see what it looks like for you.

**Review by CiceroFonsecaOSG — APPROVED** [review]
I've added some small changes:

`apps/course-authoring/.gitignore` - fixed a small typo in the comment.

`apps/course-authoring/src/renderer/renderer.d.ts` - I moved this file to the main directory to keep it close to the preload which will handle all the window stuff.

**Pull request merged**